File: back-game/src/services/roomManager.js
class RoomManager {
  constructor() {
    this.rooms = new Map(); // roomId -> { players: Map, state: {} }
    this.playerToRoom = new Map(); // playerId -> roomId
  }

  createRoom(roomId, roomData = {}) {
    if (!this.rooms.has(roomId)) {
      this.rooms.set(roomId, {
        id: roomId,
        players: new Map(),
        createdAt: Date.now(),
        ...roomData
      });
      console.log(`[RoomManager] Created room: ${roomId}`);
    }
    return this.rooms.get(roomId);
  }

  addPlayerToRoom(playerId, roomId, playerData = {}) {
    let room = this.rooms.get(roomId);
    if (!room) {
      room = this.createRoom(roomId);
    }

    room.players.set(playerId, {
      id: playerId,
      position: { x: 0, y: 0, z: 0 },
      rotation: { y: 0 },
      joinedAt: Date.now(),
      ...playerData
    });

    this.playerToRoom.set(playerId, roomId);
    console.log(`[RoomManager] Added player ${playerId} to room ${roomId}. Total: ${room.players.size}`);
    return room;
  }

  removePlayer(playerId) {
    const roomId = this.playerToRoom.get(playerId);
    if (roomId) {
      const room = this.rooms.get(roomId);
      if (room) {
        room.players.delete(playerId);
        console.log(`[RoomManager] Removed player ${playerId} from room ${roomId}. Remaining: ${room.players.size}`);

        // Delete empty room
        if (room.players.size === 0) {
          this.rooms.delete(roomId);
          console.log(`[RoomManager] Deleted empty room ${roomId}`);
        }
      }
    }
    this.playerToRoom.delete(playerId);
  }

  updatePlayerPosition(playerId, position) {
    const roomId = this.playerToRoom.get(playerId);
    if (roomId) {
      const room = this.rooms.get(roomId);
      if (room && room.players.has(playerId)) {
        room.players.get(playerId).position = position;
      }
    }
  }

  updatePlayerRotation(playerId, rotation) {
    const roomId = this.playerToRoom.get(playerId);
    if (roomId) {
      const room = this.rooms.get(roomId);
      if (room && room.players.has(playerId)) {
        room.players.get(playerId).rotation = rotation;
      }
    }
  }

  getRoom(roomId) {
    return this.rooms.get(roomId);
  }

  getRoomPlayers(roomId) {
    const room = this.rooms.get(roomId);
    if (!room) return [];
    
    return Array.from(room.players.values()).map(player => ({
      id: player.id,
      username: player.username || `player_${player.id.slice(0, 8)}`,
      position: player.position,
      rotation: player.rotation
    }));
  }

  getPlayerRoom(playerId) {
    return this.playerToRoom.get(playerId);
  }

  getAllRooms() {
    return Array.from(this.rooms.values()).map(room => ({
      id: room.id,
      playerCount: room.players.size,
      createdAt: room.createdAt
    }));
  }
}

module.exports = RoomManager;

File: back-game/services/room_manager.go
package services

import (
	"log"
	"sync"
	"time"

	"github.com/SmVynt/42trc/back-game/models"
)

type RoomManager struct {
	rooms       map[int]*models.Room
	playerRooms map[string]int // playerId -> roomId
	mu          sync.RWMutex
}

func NewRoomManager() *RoomManager {
	return &RoomManager{
		rooms:       make(map[int]*models.Room),
		playerRooms: make(map[string]int),
	}
}

func (rm *RoomManager) CreateRoom(roomID int, name string, maxPlayers int) *models.Room {
	rm.mu.Lock()
	defer rm.mu.Unlock()

	if _, exists := rm.rooms[roomID]; !exists {
		rm.rooms[roomID] = &models.Room{
			ID:         roomID,
			Name:       name,
			Players:    make(map[string]*models.Player),
			CreatedAt:  time.Now(),
			MaxPlayers: maxPlayers,
		}
		log.Printf("[RoomManager] Created room: %d", roomID)
	}
	return rm.rooms[roomID]
}

func (rm *RoomManager) AddPlayerToRoom(playerID, roomID int, playerData *models.Player) *models.Room {
	rm.mu.Lock()
	defer rm.mu.Unlock()

	// Create room if doesn't exist
	room, exists := rm.rooms[roomID]
	if !exists {
		room = &models.Room{
			ID:         roomID,
			Name:       "Room " + string(rune(roomID)),
			Players:    make(map[string]*models.Player),
			CreatedAt:  time.Now(),
			MaxPlayers: 4,
		}
		rm.rooms[roomID] = room
	}

	// Add player to room
	room.Players[playerData.ID] = playerData
	rm.playerRooms[playerData.ID] = roomID

	log.Printf("[RoomManager] Added player %s to room %d. Total: %d",
		playerData.ID, roomID, len(room.Players))

	return room
}

func (rm *RoomManager) RemovePlayer(playerID string) {
	rm.mu.Lock()
	defer rm.mu.Unlock()

	roomID, exists := rm.playerRooms[playerID]
	if !exists {
		return
	}

	room, roomExists := rm.rooms[roomID]
	if roomExists {
		delete(room.Players, playerID)
		log.Printf("[RoomManager] Removed player %s from room %d. Remaining: %d",
			playerID, roomID, len(room.Players))

		// Delete empty room
		if len(room.Players) == 0 {
			delete(rm.rooms, roomID)
			log.Printf("[RoomManager] Deleted empty room %d", roomID)
		}
	}

	delete(rm.playerRooms, playerID)
}

func (rm *RoomManager) UpdatePlayerPosition(playerID string, position models.Position) {
	rm.mu.RLock()
	defer rm.mu.RUnlock()

	roomID, exists := rm.playerRooms[playerID]
	if !exists {
		return
	}

	room, roomExists := rm.rooms[roomID]
	if roomExists && room.Players[playerID] != nil {
		room.Players[playerID].Position = position
	}
}

func (rm *RoomManager) UpdatePlayerRotation(playerID string, rotation models.Rotation) {
	rm.mu.RLock()
	defer rm.mu.RUnlock()

	roomID, exists := rm.playerRooms[playerID]
	if !exists {
		return
	}

	room, roomExists := rm.rooms[roomID]
	if roomExists && room.Players[playerID] != nil {
		room.Players[playerID].Rotation = rotation
	}
}

func (rm *RoomManager) GetRoom(roomID int) *models.Room {
	rm.mu.RLock()
	defer rm.mu.RUnlock()
	return rm.rooms[roomID]
}

func (rm *RoomManager) GetRoomPlayers(roomID int) []*models.Player {
	rm.mu.RLock()
	defer rm.mu.RUnlock()

	room, exists := rm.rooms[roomID]
	if !exists {
		return []*models.Player{}
	}

	players := make([]*models.Player, 0, len(room.Players))
	for _, player := range room.Players {
		players = append(players, player)
	}
	return players
}

func (rm *RoomManager) GetPlayerRoom(playerID string) int {
	rm.mu.RLock()
	defer rm.mu.RUnlock()
	return rm.playerRooms[playerID]
}

func (rm *RoomManager) GetAllRooms() []map[string]interface{} {
	rm.mu.RLock()
	defer rm.mu.RUnlock()

	rooms := make([]map[string]interface{}, 0)
	for _, room := range rm.rooms {
		rooms = append(rooms, map[string]interface{}{
			"id":          room.ID,
			"playerCount": len(room.Players),
			"createdAt":   room.CreatedAt,
		})
	}
	return rooms
}
